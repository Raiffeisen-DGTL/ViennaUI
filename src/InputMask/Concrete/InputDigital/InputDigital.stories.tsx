import React, { useCallback, useState } from 'react';
import { Story, Meta } from 'storybook';
import { InputDigital } from './InputDigital';

export default {
    title: 'Development/InputDigital',
    component: InputDigital,
} as Meta;

export const Overview: Story<any> = (args) => {
    const [value, setValue] = useState();
    const changeHandler = useCallback(({ value: data }) => setValue(data), []);

    return <InputDigital onChange={changeHandler} value={value} />;
};

export const ViewOnly: Story<any> = (args) => {
    return <InputDigital viewOnly value={'12312312'} />;
};

export const BugWithNullValue: Story<any> = (args) => {
    const [value, setValue] = useState();
    const changeHandler = ({ value: data }) => {
        setValue(data);
        console.log('value', data);
    };

    return <InputDigital onChange={changeHandler} value={value} />;
};
