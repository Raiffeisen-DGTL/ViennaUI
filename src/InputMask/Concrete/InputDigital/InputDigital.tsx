import React, { forwardRef } from 'react';
import { InputNumber, InputNumberProps } from '../InputNumber/InputNumber';

export type InputDigitalProps = Omit<InputNumberProps, 'scale' | 'thousandsSeparator' | 'mask'>;

export const InputDigital = forwardRef<HTMLInputElement, InputDigitalProps>((props, ref) => (
    <InputNumber ref={ref} scale={0} thousandsSeparator={''} {...props} />
));

InputDigital.displayName = 'InputDigital';
