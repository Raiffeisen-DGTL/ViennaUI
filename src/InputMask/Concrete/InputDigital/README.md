# InputDigital

## Импорт

```
import { InputDigital } from 'vienna-ui';
``` 

### Props

| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| `value` | `string \| number \| null` | — | Значение, передаваемое в компонент |
| `onChange` | `(data: { value: string \| number \| null; event: ChangeEvent<HTMLInputElement>; options: { name?: string; isComplete: boolean; unmaskedValue?: unknown; }) => void \| undefined` | — | Обработчик изменения значения |
| `parent` | `Masked<any> \| undefined` | — | Родительский маскированный объект |
| `prepare` | `(chars: string, masked: Masked<any>, flags: AppendFlagsMaskedState>) => string \| [string, ChangeDetails] \| undefined` | — | Преобразует значение перед применением маски |
| `prepareChar` | `(chars: string, masked: Masked<any>, flags: AppendFlagsMaskedState>) => string \| [string, ChangeDetails] \| undefined` | — | Преобразует **каждый символ** перед применением маски |
| `validate` | `(value: string, masked: Masked<any>, flags: AppendFlagsMaskedState>) => boolean \| undefined` | — | Проверяет, является ли значение допустимым |
| `commit` | `(value: string, masked: Masked<any>) => void \| undefined` | — | Выполняет дополнительную логику после завершения ввода |
| `format` | `(value: DateValue, masked: Masked<any>) => string` \| `(value: number, masked: Masked<any>) => string` \| `(value: string, masked: Masked<any>) => string` \| `(value: any, masked: Masked...>) => string` \| `undefined` | — | Форматирует значение в строку |
| `parse` | `(str: string, masked: Masked<any>) => DateValue` \| `(str: string, masked: Masked<any>) => number` \| `(str: string, masked: Masked<any>) => string` \| `(str: string, masked: Masked...>) => any` \| `undefined` | — | Парсит строку в typed value |
| `overwrite` | `boolean \| 'shift' \| undefined` | — | Включает режим перезаписи символов |
| `eager` | `boolean \| 'remove' \| 'append' \| undefined` | — | Режим "eager" для маски |
| `skipInvalid` | `boolean \| undefined` | — | Пропускать недопустимые символы |
| `autofix` | `boolean \| 'pad' \| undefined` | — | Автоматически исправляет значение |
| `thousandsSeparator` | `string \| undefined` | — | Разделитель тысячных (один символ) |
| `as` | `KnownTarget \| undefined` | — | Динамически изменяет отображаемый компонент или HTML-тег, например `as="input"` |
| `forwardedAs` | `KnownTarget \| undefined` | — | Используется для передачи другого тега/компонента |
| `theme` | `DefaultTheme \| undefined` | — | Тема компонента |
| `onComplete` | `(value: string, maskRef: InputMaskFactoryOpts>, e?: InputEvent \| undefined) => void \| undefined` | — | Вызывается, когда ввод завершён |
| `isFocusAllowMask` | `boolean \| undefined` | — | Позволяет фокусироваться на маске |



# InputDigital

Компонент ввода целых чисел



```
    {() => {
        const [value, setValue] = React.useState(null);
        const changeHandler = React.useCallback(({ value }) => setValue(value), []);
        return <InputDigital value={value} onChange={changeHandler} />;
    }}
```

## Внешний вид

Дизайн, размеры и другие настройки полностью идентичны компоненту Input. Для использования необходимо ознакомится с работой IMask. Подробнее о свойствах смотрите в `InputMask`;

```
    {() => {
        const [value, setValue] = React.useState(null);
        const changeHandler = React.useCallback(({ value }) => setValue(value), []);
        return (
            <Groups>
                <InputDigital value={value} onChange={changeHandler} size='xs' />
                <InputDigital value={value} onChange={changeHandler} size='s' />
                <InputDigital value={value} onChange={changeHandler} size='l' />
                <InputDigital value={value} onChange={changeHandler} size='xl' />
            </Groups>
        );
    }}
```

## Работа со строчными или числовыми значениями

Компонент помимо числовых значений может работать еще и со строчными. Для этого достаточно передать в `value` любую строку.

Под капотом компонент выполняет проверку, если передана в `value` строка, то работаем со строчными значениями и в `onChange`
будет также приходить строка. Если передано `null` или `number`, то работаем как с числом и в `onChange` будет также приходить
`null` или `number` в зависимости пустое поле или нет

Также необходимо иметь в виду, что при работе со строкой и указанными свойствами `min` и `max`, после потери фокуса значение будет проверяться
на соответствие по этим свойствам. И если передана строка с нулями в начале, например `0000100`, то при приведении к `min` или `max`
значение выведется без нулей

```
    {() => {
        const [value, setValue] = React.useState('');
        const changeHandler = React.useCallback(({ value }) => setValue(value), []);
        return <InputDigital min={150} value={value} onChange={changeHandler} />;
    }}
```

## Состояние ViewOnly

Это состояние используется, когда нужно показать значение поля без возможности изменения.
Может использоваться для построения форм, которые находятся в режиме просмотра, где все поля заполнены, но не доступны для редактирования.

Свойства:

- viewOnly - состояние `ViewOnly` (тип boolean);
- viewOnlyText - текст значения (тип ReactNode);

```
    <InputDigital viewOnly value={'12312312'} />
```