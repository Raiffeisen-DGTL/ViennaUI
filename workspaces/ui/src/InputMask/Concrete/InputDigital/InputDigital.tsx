import React from 'react';
import InputMask, { InputMaskProps } from '../../InputMask';

export const InputDigital = React.forwardRef((props: InputMaskProps, ref: React.Ref<HTMLInputElement>) => (
    <InputMask ref={ref} showMask={false} mask={/^[0-9]{0,}$/} {...props} />
));

InputDigital.displayName = 'InputDigital';
export default InputDigital;
