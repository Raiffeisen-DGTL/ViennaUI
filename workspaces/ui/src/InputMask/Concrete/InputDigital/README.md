# InputDigital

Компонент ввода произвольного набора цифр является частным случаем [`InputMask`](../../) с предустановленным значением `mask`

```jsx
const [value, setValue] = React.useState('');
const changeHandler = React.useCallback((e, data) => setValue(data.value), []);
return <InputDigital value={value} onChange={changeHandler} />;
```
