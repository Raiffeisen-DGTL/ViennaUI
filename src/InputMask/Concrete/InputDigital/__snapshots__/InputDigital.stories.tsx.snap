// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Development/InputDigital Overview smoke-test 1`] = `
<style>
  @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Regular.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Medium.woff') format('woff');
        font-weight: 500;
        font-style: normal;
    }

    @font-face {
        font-family: 'ALS Hauss';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.eot'); /* Fix IE11 */
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/ALS_Hauss-Bold.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Regular-Web.woff') format('woff');
        font-weight: normal;
        font-style: normal;
    }

    @font-face {
        font-family: 'Navigo';
        font-display: swap;
        src: url('https://www.raiffeisen.ru/common/new/images/fonts/Navigo-Bold-Web.woff') format('woff');
        font-weight: bold;
        font-style: normal;
    }
</style>
<div class="sc-bcPKhP psOLg">
  <div class="sc-jrkPcm juzgae">
    <label class="sc-iBAaJG eDDvAp">
      <div class="sc-gKHVLF hxumYj">
        <div class="sc-dkjKgF dCerbJ">
          <input aria-invalid="false"
                 class="sc-hLclGa hqQfvS"
          >
        </div>
      </div>
    </label>
  </div>
</div>
`;
